common/crypto/platform/openssl/crypto.cpp

```diff
@@ -430,6 +439,97 @@ expected::ExpectedString SignRawData(
+// (See MEN-1740) for some insight into previous issues, and the chosen fix.
+template <BIGNUM *DecoderFn(const unsigned char *, int, BIGNUM *)>
+static expected::ExpectedBytes TryASN1EncodeMenderCustomBinaryECFormat(
+	const vector<uint8_t> &signature, const mender::sha::SHA &shasum) {
+	// Verify that the marshalled keys match our expectation
+	const size_t assumed_signature_size {2 * ecdsa256keySize};
+	if (signature.size() > assumed_signature_size) {
+		return expected::unexpected(MakeError(
+			SetupError,
+			"Unexpected size of the signature for ECDSA. Expected 2*32 bytes. Got: "
+				+ to_string(signature.size())));
+	}
+	auto ecSig = unique_ptr<ECDSA_SIG, void (*)(ECDSA_SIG *)>(ECDSA_SIG_new(), ECDSA_SIG_free);
+	if (ecSig == nullptr) {
+		return expected::unexpected(MakeError(
+			SetupError,
+			"Failed to allocate the structure for the ECDSA signature" + GetOpenSSLErrorMessage()));
+	}
+
+	auto r = unique_ptr<BIGNUM, void (*)(BIGNUM *)>(
+		DecoderFn(signature.data(), ecdsa256keySize, NULL), BN_free);
+	if (r == NULL) {
+		return expected::unexpected(MakeError(
+			SetupError,
+			"Failed to extract the r(andom) part from the ECDSA signature in the binary representation"
+				+ GetOpenSSLErrorMessage()));
+	}
+	auto s = unique_ptr<BIGNUM, void (*)(BIGNUM *)>(
+		DecoderFn(signature.data() + ecdsa256keySize, ecdsa256keySize, NULL), BN_free);
+	if (s == NULL) {
+		return expected::unexpected(MakeError(
+			SetupError,
+			"Failed to extract the s(ignature) part from the ECDSA signature in the binary representation"
+				+ GetOpenSSLErrorMessage()));
+	}
+
+	// Set the r&s values in the SIG struct
+	// r & s now owned by ecSig
+	int ret {ECDSA_SIG_set0(ecSig.get(), r.release(), s.release())};
+	if (ret != OPENSSL_SUCCESS) {
+		return expected::unexpected(MakeError(
+			SetupError,
+			"Failed to set the signature parts in the ECDSA structure" + GetOpenSSLErrorMessage()));
+	}
+
+	/* Allocate some array guaranteed to hold the DER-encoded structure */
+	vector<uint8_t> arr(256);
+	unsigned char *arr_p = &arr[0];
+	int len = i2d_ECDSA_SIG(ecSig.get(), &arr_p);
+	if (len < 0) {
+		return expected::unexpected(MakeError(
+			SetupError,
+			"Failed to set the signature parts in the ECDSA structure" + GetOpenSSLErrorMessage()));
+	}
+	/* Resize to the actual size of the DER-encoded signature */
+	arr.resize(len);
+
+	return arr;
+}
+
+
+expected::ExpectedBool VerifySignData(
+	const string &public_key_path,
+	const mender::sha::SHA &shasum,
+	const vector<uint8_t> &signature);
+
+static expected::ExpectedBool VerifyECDSASignData(
+	const string &public_key_path,
+	const mender::sha::SHA &shasum,
+	const vector<uint8_t> &signature) {
+	expected::ExpectedBytes exp_der_encoded_signature =
+		TryASN1EncodeMenderCustomBinaryECFormat<BN_bin2bn>(signature, shasum)
+			.or_else([&signature, &shasum](error::Error) {
+				return TryASN1EncodeMenderCustomBinaryECFormat<BN_lebin2bn>(signature, shasum);
```

Comment:
I guess we should log the error? And maybe even with various log levels based on the actual error? (might be easier to do inside the `TryANS1...()` function)

Reply:
How about we add the error to the last `TryASN1` call, and then the receiver logs it as it sees fit? :thinking:

Reply:
If there are multiple errors, they should all end up somewhere, IMHO.